File: ops.py
import math

from tensor import Tensor
from typing import Callable

"""
Basic operations for Tensors
"""

# If two tensors are of different shapes, we need to mold them into the same shape in order to perform the operation --> "broadcasting"
def broadcast_op(t1: Tensor, t2: Tensor, combine_fn: Callable[[float], float]) -> Tensor:
    # Create the broadcasted shape
    broadcast_shape = Tensor.broadcast_shapes(t1.shape, t2.shape)

    # Compute the data w/broadcasted shape
    output_data = []

    total_size = math.prod(broadcast_shape)

    for idx in range(total_size):
        # Compute the multidimensional address in the broadcasted shape
        temp_idx = idx
        multi_idx = []
        for dim in reversed(broadcast_shape):
            multi_idx.append((temp_idx % dim,))
            temp_idx //= dim
        multi_idx.reverse()

        flat_idx_1 = tuple(multi_idx[i] if t1.shape[i] > 1 else 0 for i in range(len(multi_idx)))
        flat_idx_2 = tuple(multi_idx[i] if t2.shape[i] > 1 else 0 for i in range(len(multi_idx)))

        flat_index_1 = t1.flatten_index(flat_idx_1)
        flat_index_2 = t2.flatten_index(flat_idx_2)

        output_data.append(combine_fn(t1.data[flat_index_1], t2.data[flat_index_2]))
    
    return Tensor(output_data, broadcast_shape, requires_grad=True)

def add(t1: Tensor, t2: Tensor) -> Tensor:
    out = broadcast_op(t1, t2, lambda x, y: x + y)

    if out.requires_grad:
        # Derivative of addition operation is always one
        def grad_fn(grad_output):
            grad_t1 = grad_output
            grad_t2 = grad_output

            return grad_t1, grad_t2

        out._grad_fn = grad_fn
        out._parents = [t1, t2]

    return out

def mul(t1: Tensor, t2: Tensor) -> Tensor:
    out = broadcast_op(t1, t2, lambda x, y: x * y)

    if out.requires_grad:
        # Derivative of multiplication leaves behind all variables except the variable the derivative is w.r.t.
        def grad_fn(grad_output):
            grad_t1 = t1.data * grad_output
            grad_t2 = t2.data * grad_output

            return grad_t1, grad_t2

        out._grad_fn = grad_fn
        out._parents = [t1, t2]

    return out

def neg(t: Tensor) -> Tensor:
    out = Tensor(-t.data, t.requires_grad)

    if out.requires_grad:
        # Derivative of multiplication is always -1
        def grad_fn(grad_output):
            grad_t = -grad_output

            return grad_t

        out._grad_fn = grad_fn
        out._parents = [t]

    return out

def inv(t: Tensor) -> Tensor:
    out = Tensor(1 / t.data, t.requires_grad)

    if out.requires_grad:
        # Derivative of 1/x is -1/x^2
        def grad_fn(grad_output):
            grad_t = (-1 / (np.square(t.data))) * grad_output

            return grad_t

        out._grad_fn = grad_fn
        out._parents = [t]

    return out

def exp(t: Tensor) -> Tensor:
    t = t if isinstance(Tensor) else Tensor(t, requires_grad=True)

def eq(t1: Tensor, t2: Tensor) -> Tensor:
    comp_result = t1.data == t2.data

    return Tensor(float(comp_result))

File: ops.cpp
#include <numeric>
#include <algorithm>

#include "ops.h"
#include "tensor.h"

Tensor broadcast_op(const Tensor& t1, const Tensor& t2, float (*combine_fn)(float, float)) {
    // Create the broadcasted shape
    std::vector<int> broadcast_shape = Tensor::broadcast_shapes(t1.shape, t2.shape);

    // Compute the data with broadcasted shape
    std::vector<float> output_data;
    int total_size = std::accumulate(broadcast_shape.begin(), broadcast_shape.end(), 1, std::multiplies<float>());

    for (int i = 0; i < total_size; ++i) {
        int temp_idx = i;
        std::vector<int> multi_idx;

        // Get the multidimensional index for the one dimensional index
        for (int i = broadcast_shape.size() - 1; i >= 0; --i) {
            int dim = broadcast_shape[i];
            multi_idx.insert(multi_idx.begin(), temp_idx % dim);
            temp_idx = (int) (temp_idx / dim);
        }

        // Each tensor needs it's own flattened index to access for each index in the broadcasted space
        // If the shape is every <= 1 that means it effectively doesn't contain any information --> encode 0 for flatten
        std::vector<int> flat_idx_1;
        std::vector<int> flat_idx_2;

        for (int i = 0; i < multi_idx.size(); ++i) {
            if (t1.shape[i] > 1) {
                flat_idx_1.push_back(multi_idx[i]);
            } else {
                flat_idx_1.push_back(0);
            }

            if (t2.shape[i] > 1) {
                flat_idx_2.push_back(multi_idx[i]);
            } else {
                flat_idx_2.push_back(0);
            }
        }

        int flat_index_1 = t1.flatten_index(flat_idx_1);
        int flat_index_2 = t2.flatten_index(flat_idx_2);

        // Perform the elementwise operation
        output_data.push_back(combine_fn(t1.data[flat_index_1], t2.data[flat_index_2]));
    }

    return Tensor(output_data, broadcast_shape, true);
}

Tensor add(Tensor t1, Tensor t2) {
    Tensor output_t = broadcast_op(t1, t2, [](float a, float b) { return a + b; });

    auto grad_fn = [](const Tensor& upstream_grad) {
        return std::vector<Tensor>{upstream_grad, upstream_grad};
    };

    output_t.grad_fn = grad_fn;
    output_t.parents = {t1, t2};

    return output_t;
}

Tensor mul(Tensor t1, Tensor t2) {
    auto mul = [](float a, float b) { return a * b; };
    Tensor output_t = broadcast_op(t1, t2, mul);

    auto grad_fn = [&t1, &t2, &mul](const Tensor& upstream_grad) {
        return std::vector<Tensor>{
            broadcast_op(t2, upstream_grad, mul),
            broadcast_op(t1, upstream_grad, mul)
        };
    };

    output_t.grad_fn = grad_fn;
    output_t.parents = {t1, t2};

    return output_t;
}

Tensor neg(Tensor t) {
    auto neg = [](float a) { return -a; };

    Tensor output_t(t.data, t.shape, true);
    std::transform(output_t.data.begin(), output_t.data.end(), output_t.data.begin(), neg);

    auto grad_fn = [&neg](const Tensor& upstream_grad) {
        Tensor neg_upstream = upstream_grad;
        std::transform(neg_upstream.data.begin(), neg_upstream.data.end(), neg_upstream.data.begin(), neg);

        return std::vector<Tensor>{neg_upstream};
    };

    output_t.grad_fn = grad_fn;
    output_t.parents = std::vector<Tensor>{t};

    return output_t;
}

Tensor inv(Tensor t) {
    auto inv = [](float a) { return 1 / a; };

    Tensor output_t(t.data, t.shape, true);
    std::transform(output_t.data.begin(), output_t.data.end(), output_t.data.begin(), inv);

    auto grad_fn = [](const Tensor& upstream_grad) {
        auto inv_deriv = [](float a) { return -1 / (a * a); };

        Tensor inv_upstream = upstream_grad;
        std::transform(inv_upstream.data.begin(), inv_upstream.data.end(), inv_upstream.data.begin(), inv_deriv);
        
        return std::vector<Tensor>{inv_upstream};
    };

    output_t.grad_fn = grad_fn;
    output_t.parents = std::vector<Tensor>{t};

    return output_t;
}